File: dsa/py/basics/sorting/bubble_sort.py
"""
Swap locally; Bubble / Move the largest element to the end

T: O(n^2)
S: O(1)
"""


def sort(arr):
    n = len(arr)

    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                # swap
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
        print(f"partial sort -> {arr}")

    return arr


print(sort([3, 4, 1, 2, 5]))

File: dsa/java/DSA/Concepts/Sorting/bubble_sort.java
package Algorithms.Sorting;

import Common.io;

public class bubble_sort {
    public static void main(String[] args) {
        int[] arr = io.readArr();
        System.out.print("Before Sort = ");
        io.displayArr(arr);

        // Algo
        // Swap locally
        // Bubble / Move the largest element to the end

        int n = arr.length;

        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    // swap
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }

        System.out.print("After Sort = ");
        io.displayArr(arr);
    }
}
